### DISCUSSION TITLE
[wasm] _malloc and _free functions are not exported from zxing_reader.js?

### DISCUSSION THREAD

**Discussion by dkonecny-oracle** [zxing-cpp/zxing-cpp#510] [Q&A] [answered]
I followed https://github.com/zxing-cpp/zxing-cpp/tree/master/wrappers/wasm and compiled it and got it to work. One thing I stumbled over was that the demo_reader.html is calling zxing._malloc and zxing._free, but these two methods are not exported in zxing_reader.js which I compiled. The method is declared only for internal scope, ie.:

            var _malloc = function() {
                return (_malloc = Module["asm"]["ka"]).apply(null, arguments)
            };

 While in the zxing_reader.js in your live demo it is exported, ie:

            var _malloc = Module["_malloc"] = function() {
                return (_malloc = Module["asm"]["ka"]).apply(null, arguments)
            };

What do I need to do to get these two functions exported automatically so that the demo reader works and I do not have to manually fix the generated code?
Thanks,
-David

**Comment by axxel** [comment, marked as answer]
I could not reproduce your issue (but actually found another one, see 72023ad3dded9f3f94b5d85bc10385e0cf76620a). Googling for your problem led me [here](https://github.com/emscripten-core/emscripten/issues/6882). But here the demo_reader just works, including building it with `-O3`. If you find some insight there, please let me know.

**Comment by dkonecny-oracle** [comment]
Thx Axxel, the emscripten link helped me resolve this. I simply explicitly enumerated _malloc and _free as functions to export as both are needed and both were missing and that resolve my issue. Here is the patch:
[a.patch](https://github.com/zxing-cpp/zxing-cpp/files/10679632/a.patch)

**Reply by axxel** [reply]
Thanks for that. I know now what the difference between your build on one hand and mine + the CI build on the other is: the emsdk version. After I updated to the latest, I get the same error. I will commit your patch.
